### DISCUSSION TITLE
shutdown all servers before starting a new one

### DISCUSSION THREAD

**Discussion by lolbinarycat** [joaotavora/eglot#1452] [Q&A] [answered]
My laptop doesn't have much ram, and while it can just about run one language server, if i ever start 2-3 lang servers, it will potentially lead to emacs getting oom killed.

I found `eglot-shutdown-all`, but I can't find a hook that is run before starting up a new lang server.  `eglot-mode-hook` also doesn't seem to exist.

EDIT: `eglot-server-initialized-hook` is the correct hook.  `eglot-connect-hook` will not work!

**Comment by joaotavora** [comment]
eglot-connect-hook maybe

**Reply by lolbinarycat** [reply]
eglot-connect-hook weirdly seems to expect its functions to have an argument, which isn't documented anywhere...

**Reply by joaotavora** [reply]
Why would it be weird for a hook called when connecting to a given server to take that server as an argument? But you're right there's no doc for that

**Reply by joaotavora** [reply]
Anyway, does it do what you want?

**Reply by lolbinarycat** [reply]
yeah, but only if the lambda takes and ignores an argument.

generally having to reverse engineer stuff is bad ux.

**Reply by joaotavora** [reply]
so sorry you had such a bad ux solving your problem with my advice.  I'll be sure to abstain from helping you in the future

**Reply by lolbinarycat** [reply]
I found `eglot-connect-hook` on my own, read the doc comment that says it is "run after connecting".  that seems to pretty unambiguously say that it is not the right hook for my usecase.  but somehow i'm wrong, actually, and "before" actually means "after".

i then tried your suggestion, and it didn't work.

i then read the function signature of other connection hooks, and reverse engineered the problem from there.

i have half a mind to submit a PR with a more helpful doc comment, but now i'm not sure, since you don't seem to appreciate constructive criticism.

**Reply by joaotavora** [reply]
>  but somehow i'm wrong, actually, and "before" actually means "after".

What are you on about?  eglot-connect-hook's doc says it is run "after" connecting.   And that's what it does, period. The only thing missing there is a note that it takes the server just connected to as a hook argument. If you shut down other servers using `eglot-shutdown` in that hook, it does what you want I think.

And possibly `eglot-server-initialized-hook` also works.  And that hook's docstring is accurate.

> i then tried your suggestion, and it didn't work.

Then why on earth did you say it did after I asked you?  What relevance is that comment about "weirdly takes an argument"?  It's not weird at all.

Anyway, your rather bizarre use case is probably solved easier by not using `eglot-ensure` (since you're such a doc and ux buff you have read the docs about that) or  just remembering to to shutdown before starting `M-x eglot`.

**Reply by lolbinarycat** [reply]
my point was it didn't work *initially*, only after i did other stuff.

i think the old issue about the pre-connect hook should be revived.

**Reply by joaotavora** [reply]
There is already a pre-connect hook.

**Comment by joaotavora** [comment, marked as answer]
Don't use `eglot-ensure`.  Shutdown existing servers before starting new ones.

**Reply by lolbinarycat** [reply]
i'm already not using eglot-ensure.

this isn't an answer, this is just repeating my question.

**Reply by joaotavora** [reply]
If you're not using eglot-ensure you're using M-x eglot. If this breaks down after 3 servers just you n have to learn to count to 3.

Or make some `M-x lolglot` command that shuts things down and then calls `eglot` interactively. 0  reverse engineering there.
